### PR TITLE
Initial contribution from SVRWB benchmark (take 2)

### PR DESCRIPTION
**Pull request by roachspray**
The original pull request was #214 , but due to some issues with an app was canceled and re-spun as this pr:

- This pull request is related to issue #50 and includes:
  - `jasper-1.701.0`
  - `perl-5.21.7`
      - This case is kind of interesting and we should be careful because it uses it's own `Configure` shell script, instead of `configure`.
  - `tcpdump-4.9.0`
- For each of the `_fuzz.cc` harness files also has a main() routine which is included upon ifdef `_HAS_MAIN`.
- For each of these, a single seed is given that was randomly chosen.
- For each of these, a subdirectory named `known_bugs` exists that contains a number of *known* bugs. 
- Currently, there is no test case for checking inputs in `known_bugs`, but that will be forthcoming in a future pr.
- These have passed `make build-afl-$BENCHMARK_NAME` and `make run-afl-$BENCHMARK_NAME` on my setup.

There is some debate to the benefit of these bugs (see #165 for deeper discussion), but it perhaps can act as a starting place for getting other pieces of comparison/measurement in working order.

If these are accepted, I will plan to look to add some of the others from the svrwb set.

This work is done with the support of the Veracode Applied Research Group.

cc @jonathanmetzman  @lszekeres

### CODE REVIEW

**Comment by roachspray** [comment]
I see the conflicts and they are no issue -- I am testing them now with them merged...

**Comment by roachspray** [comment]
Ok, this should be good to go (pending CI) .. but my local build-$FUZZER-$BENCH and run-$FUZZER-$BENCH passed for the 3 I have added. I had to re-work the tcpdump case because of dependencies and the desire to have them isolated to the benchmark; might be an OK example for other's to reference if they need to do something similar. 

...I think that's it.

**Comment by roachspray** [comment]
Anyone with any clue on `eclipser` failures?

**Comment by lszekeres** [comment]
> Anyone with any clue on `eclipser` failures?

AFAIK it is unrelated to your PR, so please don't worry about it.

File: benchmarks/jasper-1.701.0/build.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# XXX There is no hash to check. Git repo for this only goes back to 1.9x.
+#
+ZIPURL=https://www.ece.uvic.ca/~frodo/jasper/software/jasper-1.701.0.zip
+rm -rf SRC jasper-1.701.0 jasper-1.701.0.zip
+wget ${ZIPURL} || (printf "wget failed" && exit 1)
+unzip jasper-1.701.0.zip || (printf "unzip failed" && exit 1)
+mv jasper-1.701.0 SRC
+
+build_lib
+
```
Comment (jonathanmetzman) [inline benchmarks/jasper-1.701.0/build.sh:34]: nit: remove this empty line

File: benchmarks/tcpdump-4.9.0/build.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+. $(dirname $0)/../common.sh
+
+build_lib() {
+  rm -rf BUILD
+  cp -rf SRC BUILD
+  # we build and install bc was having issues with autogen.sh of dbus
+  # makes building the rest easier too..
```
Comment (jonathanmetzman) [inline benchmarks/tcpdump-4.9.0/build.sh:22]: nit one period.
Also capitalize the beginning of the sentence please.

File: benchmarks/tcpdump-4.9.0/build.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   make install)
+  (cd BUILD/libdbus && 
+    CFLAGS="${CFLAGS} -Wno-error" ./autogen.sh --disable-xml-docs --disable-doxygen-docs --disable-ducktype-docs &&
+    make && make install)
+  (cd BUILD/libpcap && ./configure && make &&
+   cd ../tcpdump && ./configure && make)
+}
+
+#
+# tcpdump really wants libpcap to share the same parent directory.
```
Comment (jonathanmetzman) [inline benchmarks/tcpdump-4.9.0/build.sh:33]: Can you please drop the empty comment lines below and above this line?

File: benchmarks/tcpdump-4.9.0/build.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+# 
+mkdir SRC
+git clone https://github.com/libexpat/libexpat.git SRC/libexpat
+get_git_tag https://gitlab.freedesktop.org/dbus/dbus.git  dbus-1.10 SRC/libdbus
+get_git_tag https://github.com/the-tcpdump-group/libpcap.git  libpcap-1.9.0 SRC/libpcap
+get_git_tag https://github.com/the-tcpdump-group/tcpdump.git  tcpdump-4.9.0 SRC/tcpdump
+
+build_lib
+
+# copy over shared lib deps to /out
```
Comment (jonathanmetzman) [inline benchmarks/tcpdump-4.9.0/build.sh:43]: nit: capitalize the beginning of this sentence and end it with a period please.

File: benchmarks/tcpdump-4.9.0/build.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+cp BUILD/libexpat/expat/lib/.libs/libexpat.so.1.6.11 $OUT/
+ln -s $OUT/libexpat.so.1.6.11 $OUT/libexpat.so.1
+ln -s $OUT/libexpat.so.1.6.11 $OUT/libexpat.so
+cp BUILD/libdbus/dbus/.libs/libdbus-1.so.3.14.16 $OUT/
+ln -s $OUT/libdbus-1.so.3.14.16 $OUT/libdbus-1.so.3
+ln -s $OUT/libdbus-1.so.3.14.16 $OUT/libdbus-1.so
+
+#
+# To test with the main() in tcpdump_fuzz.cc, use -D_HAS_MAIN and disable any
```
Comment (jonathanmetzman) [inline benchmarks/tcpdump-4.9.0/build.sh:52]: Can you please drop the empty comment lines below and above this line?

File: docker/base-builder/Dockerfile [hunk 6]
```diff
@@ -21,9 +21,13 @@ FROM gcr.io/fuzzbench/base-image
 # recommended for some reason.
 RUN apt-get update -y && apt-get install -y --no-install-recommends \
     automake \
+    bison \
```
Comment (jonathanmetzman) [inline docker/base-builder/Dockerfile:24]: I've just stopped installing dependencies in this file unless it is very generally useful.
cmake and unzip might be appropriate here but I'd rather everything gets installed in build.sh.
Sorry to change things on you in the middle.
Here's an example of how I do this: https://github.com/google/fuzzbench/blob/master/benchmarks/freetype2-2017/build.sh#L18
(we don't have sudo in the images so the script assumes running as root, Do you need these to run on your host?)

File: benchmarks/tcpdump-4.9.0/tcpdump_fuzz.cc [hunk 7]
```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 Google LLC
```
Comment (jonathanmetzman) [inline benchmarks/tcpdump-4.9.0/tcpdump_fuzz.cc:1]: We don't have an official style for C/C++ but I may clang-format this in the future.
No need to change anything now though.

File: benchmarks/perl-5.21.7/build.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+#     -Alddlflags="-shared"
+#
+  (cd BUILD && ./Configure -des -Dusedevel && make)
+}
+
+get_git_tag https://github.com/Perl/perl5.git v5.21.7 SRC
+build_lib
+
+#
+# To test with the main() in perl_fuzz.cc, use -D_HAS_MAIN and disable any
```
Comment (jonathanmetzman) [inline benchmarks/perl-5.21.7/build.sh:40]: Can you drop the empty comment lines please?

File: benchmarks/jasper-1.701.0/build.sh [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+#
+# To test with the main() in jasper_fuzz.cc, use -D_HAS_MAIN and disable any
```
Comment (jonathanmetzman) [inline benchmarks/jasper-1.701.0/build.sh:37]: nit: can you remove the empty comment lines.

File: benchmarks/jasper-1.701.0/build.sh [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+. $(dirname $0)/../common.sh
+
+build_lib() {
+  rm -rf BUILD
+  cp -rf SRC BUILD
+  (cd BUILD && ./configure --enable-static --disable-shared && make)
+}
+
+#
+# XXX There is no hash to check. Git repo for this only goes back to 1.9x.
```
Comment (jonathanmetzman) [inline benchmarks/jasper-1.701.0/build.sh:25]: nit: can you remove the empty comment lines.

**Comment by roachspray** [comment]
@jonathanmetzman Thank you! I think I have addressed the issues you have mentioned.

File: docker/base-builder/Dockerfile (hunk 6 above)
Comment (roachspray) [reply docker/base-builder/Dockerfile]: Fixed.

File: benchmarks/jasper-1.701.0/build.sh (hunk 1 above)
Comment (roachspray) [reply benchmarks/jasper-1.701.0/build.sh:32]: Fixed

File: benchmarks/jasper-1.701.0/build.sh (hunk 9 above)
Comment (roachspray) [reply benchmarks/jasper-1.701.0/build.sh:33]: Fixed

File: benchmarks/tcpdump-4.9.0/build.sh (hunk 5 above)
Comment (roachspray) [reply benchmarks/tcpdump-4.9.0/build.sh:48]: Fixed

File: benchmarks/tcpdump-4.9.0/build.sh (hunk 4 above)
Comment (roachspray) [reply benchmarks/tcpdump-4.9.0/build.sh]: Fixed.

File: benchmarks/tcpdump-4.9.0/build.sh (hunk 3 above)
Comment (roachspray) [reply benchmarks/tcpdump-4.9.0/build.sh:32]: Fixed

File: benchmarks/tcpdump-4.9.0/build.sh (hunk 2 above)
Comment (roachspray) [reply benchmarks/tcpdump-4.9.0/build.sh]: Fixed

File: benchmarks/jasper-1.701.0/build.sh (hunk 1 above)
Comment (roachspray) [reply benchmarks/jasper-1.701.0/build.sh:32]: Fixed

**Comment by roachspray** [comment]
I am uncertain why some are checks are failing. Looking at logs, it was not obvious to me. In one case I saw earlier, there were issues with git clone in one of the other benchmarks, but looked random occurrence.

**Comment by jonathanmetzman** [comment]
I'll land this on Monday so if there are any breakages we somehow missed, they can be fixed on a workday.
Thank you for this contribution!

**Comment by roachspray** [comment]
> I'll land this on Monday so if there are any breakages we somehow missed, they can be fixed on a workday.

Thank you.

In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). Should I change now or after this lands?

**Comment by roachspray** [comment]
Prior to merge, I have another concern regarding exit():

Deeper in perl and tcpdump, `exit()` might still be called and `abort()` deeper in jasper. This is not ideal/proper, for obvious reasons, but modification to make these OK may be non-trivial.

Should this be considered a best-effort or that we must root all these out?

Is it true all the existing cases (not these new ones) all do not invoke exit() (or another bad undesirable function) somewhere in the library code?

**Comment by jonathanmetzman** [comment]
> > I'll land this on Monday so if there are any breakages we somehow missed, they can be fixed on a workday.
> 
> Thank you.
> 
> In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). I have pushed these changes and they work locally (`build-afl-*` and `run-afl-*`).
> 
> Another note: the jasper seed, a JPEG-2000 file, was crashing for me on the initial run ... which it was not prior to adding to git and CI was not seeing that, I don't think. I have replaced it with another. Oddly, my first attempt to replace this crashing seed by generating one from Preview.app also crashed ... :-/ Anyway, it is replaced.



> > I'll land this on Monday so if there are any breakages we somehow missed, they can be fixed on a workday.
> 
> Thank you.
> 
> In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). I have pushed these changes and they work locally (`build-afl-*` and `run-afl-*`).

Oh this is not good. Maybe the nonzero paths weren't being reached? Can we make it so that all paths through program return zero?
Which benchmarks do this, I didn't notice this when I reviewed (except perl benchmark I see now can return nonzero).

> 
> Another note: the jasper seed, a JPEG-2000 file, was crashing for me on the initial run ... which it was not prior to adding to git and CI was not seeing that, I don't think. I have replaced it with another. Oddly, my first attempt to replace this crashing seed by generating one from Preview.app also crashed ... :-/ Anyway, it is replaced.

Also weird, can you share the seed? I can try it out.

**Comment by roachspray** [comment]
> > In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). I have pushed these changes and they work locally (`build-afl-*` and `run-afl-*`).
> 
> Oh this is not good. Maybe the nonzero paths weren't being reached? Can we make it so that all paths through program return zero?
> Which benchmarks do this, I didn't notice this when I reviewed (except perl benchmark I see now can return nonzero).
> 

I think so. 

 Question: what if we caused a crash at those points? The crashes could be removed as known and would lessen changes on target code base.

**Comment by jonathanmetzman** [comment]
> Prior to merge, I have another concern regarding exit():
> 
> Deeper in perl and tcpdump, `exit()` might still be called and `abort()` deeper in jasper. This is not ideal/proper, for obvious reasons, but modification to make these OK may be non-trivial.
> 
> Should this be considered a best-effort or that we must root all these out?

Hmm...I don't really know the answer. libFuzzer normally is incompatible with these. But because of the weird way we run libFuzzer (-fork=1 -ignore_crashes=1) it may actually be OK if not ideal. This is probably why CI didn't catch this issue. I may need to think about this more to decide if it is OK, but it would probably be easier for me if they were removed. One issue that may come up is we currently measure the coverage of the entire AFL output directory including crashes. Since our coverage measurement is done using libFuzzer, it may be slowed down by having a lot of crashes.

Locally I added a somewhat shallow exit/abort into https://raw.githubusercontent.com/llvm-mirror/compiler-rt/master/test/fuzzer/MultipleConstraintsOnSmallInputTest.cpp. It slows down libFuzzer and AFL but both seem to be able to find new coverage. 
CC @kcc and @morehouse to see if they think it makes sense to benchmark libFuzzer under these conditions.

> 
> Is it true all the existing cases (not these new ones) all do not invoke exit() (or another bad undesirable function) somewhere in the code? Obviously, most stand alone libraries likely should not be invoking exit, as is, but...right.. real world.

I'm pretty sure none of them invoke exit.

**Comment by jonathanmetzman** [comment]
> > > In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). I have pushed these changes and they work locally (`build-afl-*` and `run-afl-*`).
> > 
> > 
> > Oh this is not good. Maybe the nonzero paths weren't being reached? Can we make it so that all paths through program return zero?
> > Which benchmarks do this, I didn't notice this when I reviewed (except perl benchmark I see now can return nonzero).
> 
> I think so.
> 
> Question: what if we caused a crash at those points in the target library that do use exit() or another function that is undesirable? The crashes could be removed as known and would lessen changes on target code base.

Sorry I didn't see this before my last comment. I think exit and abort are equivalent to crashes (returning nonzero should be removed though). My hunch is libFuzzer in our setup can work despite crashes although it is hurt by them. One issue is that this might not be fair to libFuzzer (though some would argue that it is). But a bigger issue is that this could down measuring. We may just need to try it to find out what happens :-)

**Comment by roachspray** [comment]
> > > > In going back through the documentation, one thing that is off with my harnesses is not always returning 0 from `LLVMFuzzerTestOneInput` and, in fact, calling abort() and exit() from there (in some cases). I have pushed these changes and they work locally (`build-afl-*` and `run-afl-*`).
> > > 
> > > 
> > > Oh this is not good. Maybe the nonzero paths weren't being reached? Can we make it so that all paths through program return zero?
> > > Which benchmarks do this, I didn't notice this when I reviewed (except perl benchmark I see now can return nonzero).
> > 
> > 
> > I think so.
> > Question: what if we caused a crash at those points in the target library that do use exit() or another function that is undesirable? The crashes could be removed as known and would lessen changes on target code base.
> 
> Sorry I didn't see this before my last comment. I think exit and abort are equivalent to crashes (returning nonzero should be removed though). My hunch is libFuzzer in our setup can work despite crashes although it is hurt by them. One issue is that this might not be fair to libFuzzer (though some would argue that it is). But a bigger issue is that this could down measuring. We may just need to try it to find out what happens :-)

That works for me... :) it can always be turned off if it's clear there are changes that should be made.

**Comment by jonathanmetzman** [comment]
I still need to figure what the next experiment will be, it will either be the new fuzzers or all fuzzers on the new benchmarks, but I hope to run this soon.

**Comment by jonathanmetzman** [comment]
I started the experiment (see https://github.com/google/fuzzbench/issues/249#issuecomment-623802265) for details. Sorry it took a while

**Comment by roachspray** [comment]
No worries whatsoever... thank you for the heads up!

**Pull request merged**